cpp/include/cudf/detail/join/filtered_join.cuh

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/join/join.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/primitive_row_operators.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cuco/bucket_storage.cuh>
+#include <cuco/extent.cuh>
+#include <cuco/static_multiset_ref.cuh>
+#include <cuco/static_set_ref.cuh>
+#include <cuco/types.cuh>
+#include <cuda/std/type_traits>
+
+#include <type_traits>
+
+// Forward declaration
+namespace cudf::experimental::row::equality {
+class preprocessed_table;
+}
+
+namespace cudf {
+namespace detail {
+
+using cudf::experimental::row::lhs_index_type;
+using cudf::experimental::row::rhs_index_type;
+
+/**
+ * @brief Base class providing common functionality for filtered join operations.
+ *
+ * This abstract class implements the core components needed for hash-based semi
+ * and anti join operations.
+ */
+class filtered_join {
+ public:
+  /**
+   * @brief Properties of the build table used in the join operation
+   */
+  struct build_properties {
```

Comment:
Can this be used not just for the "build" table, such as the "probe" table? If so, we should instead call it `table_property` or similar.